src/core/or/connection_edge.c

```diff
@@ -863,18 +863,34 @@ export_hs_client_circuit_id(edge_connection_t *edge_conn,
   if (protocol != HS_CIRCUIT_ID_PROTOCOL_HAPROXY)
     return;
 
+  const origin_circuit_t *circ = NULL;
   char *buf = NULL;
   const char dst_ipv6[] = "::1";
   /* See RFC4193 regarding fc00::/7 */
-  const char src_ipv6_prefix[] = "fc00:dead:beef:4dad:";
+  const char src_ipv6_prefix[] = "fc00:";
   uint16_t dst_port = 0;
   uint16_t src_port = 1; /* default value */
   uint32_t gid = 0; /* default value */
+  char *rp_identity_buf = NULL;
 
   /* Generate a GID and source port for this client */
   if (edge_conn->on_circuit != NULL) {
-    gid = TO_ORIGIN_CIRCUIT(edge_conn->on_circuit)->global_identifier;
+    circ = CONST_TO_ORIGIN_CIRCUIT(edge_conn->on_circuit);
+    gid = circ->global_identifier;
     src_port = gid & 0x0000ffff;
+    if (get_options()->HiddenServiceExportRendPoint) {
+      extend_info_t *extend_info = circ->build_state->chosen_exit;
+      if (BUG(extend_info == NULL)) {
+        goto cleanup;
+      }
+      char *rp_identity = (char *) hex_str(extend_info->identity_digest,
+                                           DIGEST_LEN);
+      tor_strlower(rp_identity);
+      tor_asprintf(&rp_identity_buf, "%.4s:%.4s:%.4s",
+                   rp_identity,
+                   rp_identity + 4,
+                   rp_identity + 8);
+    }
   }
 
   /* Grab the original dest port from the hs ident */
```

Comment:
Here's another undocumented feature: the destination port is the virtual port from the onion service protocol.

Reply:
It will need tests, too :-)

Reply:
It will need tests, too :-)

Reply:
@ahf: whatever we decide to do with the source port, we should do the same thing with the destination port. I think we should just document them, to avoid breaking existing setups.

Reply:
Documented in 49127e7.